File: flutter/lib/src/connection.dart
import 'dart:async';
import 'dart:convert';

import 'api.dart';
import 'backoff.dart';
import 'cancel.dart';
import 'models.dart';
import 'options.dart';
import 'protocol.dart';
import 'socket.dart';
import 'version.dart';

/// One multiplexed socket to the gateway, with the reconnect behaviour a shipped app needs.
///
/// Reconnecting is not a nicety here. A gateway is stateful, so a rolling update drops every
/// socket on the node being replaced, and all of those clients come back at the same instant.
/// Delays are drawn with **full jitter** — uniformly from `0..ceiling`, ceiling doubling to 30s —
/// because that is what actually de-synchronises a fleet. A fixed backoff, or one jittered by
/// ±10%, reconnects a whole node's worth of clients inside the same second, against the service
/// that has just come back up.
///
/// 网关一次滚动更新会让该节点上所有客户端同时重连。退避必须是全抖动（0..上限 均匀取值），
/// 固定退避或 ±10% 抖动会让它们挤在同一秒里打向刚恢复的服务。
class ImConnection implements ImRequester {
  ImConnection(this._options) : _token = _options.token;

  ImOptions _options;
  String _token;

  final FullJitterBackoff _backoff = FullJitterBackoff();

  final Map<String, Completer<ImFrame>> _pending = <String, Completer<ImFrame>>{};
  final Map<String, List<void Function(ImFrame)>> _listeners =
      <String, List<void Function(ImFrame)>>{};

  final StreamController<ImConnectionState> _states =
      StreamController<ImConnectionState>.broadcast();
  final StreamController<ImKickEvent> _kicks = StreamController<ImKickEvent>.broadcast();

  ImSocket? _socket;
  StreamSubscription<dynamic>? _subscription;
  Timer? _reconnectTimer;
  Timer? _heartbeatTimer;
  Future<bool>? _reauthing;

  ImConnectionState _state = ImConnectionState.idle;
  Duration _heartbeatInterval = const Duration(seconds: 30);
  int _attempt = 0;
  int _sequence = 0;
  bool _stopped = false;

  ImConnectionState get state => _state;

  /// Connection state transitions, for a status banner.
  Stream<ImConnectionState> get states => _states.stream;

  /// Emitted when the server ends the session rather than the network doing it.
  Stream<ImKickEvent> get kicks => _kicks.stream;

  /// The token currently in use, which is not `options.token` after a `conn.reauth`.
  String get token => _token;

  /// How many replies are outstanding. Exposed because "cancellation leaks a pending entry" is a
  /// bug you can only see from here, and it is one of the conformance tests.
  int get pendingRequestCount => _pending.length;

  /// Subscribes to a server push target. Returns a function that removes the subscription.
  void Function() on(String target, void Function(ImFrame frame) listener) {
    _listeners.putIfAbsent(target, () => <void Function(ImFrame)>[]).add(listener);
    return () => _listeners[target]?.remove(listener);
  }

  Future<void> connect() async {
    _stopped = false;
    await _open();
  }

  Future<void> close() async {
    _stopped = true;
    _clearTimers();
    await _subscription?.cancel();
    _subscription = null;
    await _socket?.close(1000, 'client-close');
    _socket = null;
    _setState(ImConnectionState.closed);
    _failAllPending();
  }

  /// Sends a request and returns its business payload, throwing [ImException] on a non-zero code.
  ///
  /// Requests issued while offline fail immediately rather than queueing: a chat client that
  /// silently buffers sends produces messages that arrive minutes later with no explanation, and
  /// the user has already retyped them by then. The application knows whether a message is still
  /// worth sending; the SDK does not.
  ///
  /// A reply carrying [ImErrorCode.tokenExpired] is retried **once**, after a `conn.reauth` on the
  /// socket that is already open — see [ImOptions.onTokenExpired]. Only if that reauth fails does
  /// the error reach the caller.
  @override
  Future<T> request<T>(String target, [Object? body, ImCancelToken? cancel]) async {
    final Object? data = await _requestData(target, body, cancel, allowReauth: true);
    return data as T;
  }

  /// Sends a request and returns the raw frame, for callers that need the envelope.
  Future<ImFrame> send(String target, [Object? body, ImCancelToken? cancel]) {
    if (cancel != null && cancel.isCancelled) {
      return Future<ImFrame>.error(
        ImCancelledException(target: target, reason: cancel.reason),
      );
    }

    final ImSocket? socket = _socket;
    if (socket == null || _state != ImConnectionState.open) {
      // 1005 ServiceUnavailable, not 1004: nothing was delivered, and the caller can say so.
      return Future<ImFrame>.error(
        const ImException(ImErrorCode.serviceUnavailable, 'not connected'),
      );
    }

    final String id = 'c${++_sequence}-${DateTime.now().microsecondsSinceEpoch.toRadixString(36)}';
    final Completer<ImFrame> completer = Completer<ImFrame>();
    _pending[id] = completer;

    final Timer timer = Timer(_options.requestTimeout, () {
      if (_pending.remove(id) != null && !completer.isCompleted) {
        completer.completeError(
          ImException(ImErrorCode.timeout, 'request timed out: $target', target: target),
        );
      }
    });

    // Cancelling must remove the pending entry, not merely stop the caller waiting: an entry left
    // behind is a leak that grows for the life of the connection.
    final void Function()? unlisten = cancel?.addListener(() {
      if (_pending.remove(id) != null && !completer.isCompleted) {
        completer.completeError(
          ImCancelledException(target: target, reason: cancel.reason),
        );
      }
    });

    void cleanup() {
      timer.cancel();
      unlisten?.call();
    }

    try {
      socket.send(jsonEncode(
        ImRequest(id: id, target: target, body: body ?? const <String, dynamic>{}).toJson(),
      ));
    } catch (error) {
      cleanup();
      _pending.remove(id);
      return Future<ImFrame>.error(
        ImException(ImErrorCode.serviceUnavailable, 'send failed: $error', target: target),
      );
    }

    return completer.future.whenComplete(cleanup);
  }

  // -------------------------------------------------------------------------- internals

  Future<Object?> _requestData(
    String target,
    Object? body,
    ImCancelToken? cancel, {
    required bool allowReauth,
  }) async {
    final ImFrame frame = await send(target, body, cancel);
    final ImException? error = ImException.fromFrame(frame, target);
    if (error == null) return frame.body?.data;

    // §7.4: an expired token costs one frame, not a reconnect. Retried exactly once — a second
    // expiry on a token minted seconds ago is a server or clock problem, and looping on it turns
    // one failure into a storm.
    if (allowReauth &&
        error.code == ImErrorCode.tokenExpired &&
        _options.onTokenExpired != null &&
        await _reauth()) {
      return _requestData(target, body, cancel, allowReauth: false);
    }

    throw error;
  }

  /// Renews the token on the live socket. Concurrent callers share one attempt — a burst of
  /// requests all failing 1101 at once must produce one `conn.reauth`, not one each.
  Future<bool> _reauth() {
    return _reauthing ??= _performReauth().whenComplete(() => _reauthing = null);
  }

  Future<bool> _performReauth() async {
    final String? next = await _options.onTokenExpired?.call();
    if (next == null || next.isEmpty) return false;

    try {
      await _requestData(
        'conn.reauth',
        <String, Object?>{'token': next},
        null,
        allowReauth: false,
      );
    } catch (_) {
      // The session is genuinely finished; let the original error surface to the caller and let
      // the close path handle the drop.
      return false;
    }

    _token = next;
    _options = _options.copyWith(token: next);
    return true;
  }

  Future<void> _open() async {
    _setState(_attempt == 0 ? ImConnectionState.connecting : ImConnectionState.reconnecting);

    final ImSocket socket = (_options.socketFactory ?? defaultImSocketFactory)(_buildUri());
    _socket = socket;

    _subscription = socket.messages.listen(
      _dispatch,
      onError: (Object _) {
        // onDone always follows, and the reconnect decision lives there so it runs exactly once.
      },
      onDone: () => _handleClosed(socket),
      cancelOnError: false,
    );

    try {
      await socket.ready;
    } catch (_) {
      // A refused handshake never produces a close frame, so nothing else would notice it.
      _handleClosed(socket);
      return;
    }

    _attempt = 0;
    _setState(ImConnectionState.open);
    _startHeartbeat();
  }

  Uri _buildUri() {
    final String base = _options.endpoint.replaceAll(RegExp(r'/+$'), '');
    final Map<String, String> query = <String, String>{
      'appId': _options.appId,
      'token': _token,
      'deviceId': _options.deviceId,
      'platform': _options.platform.wireValue.toString(),
      'v': '1',
      // The SDK's own version when the host app did not name one, so a support ticket carries it
      // without anyone having to ask which build the customer is on.
      'cv': _options.clientVersion ?? ImSdk.packageVersion,
      if (_options.language != null) 'lang': _options.language!,
    };

    return Uri.parse('$base${_options.channel}').replace(queryParameters: query);
  }

  void _dispatch(dynamic raw) {
    if (raw is! String) return;

    final ImFrame frame;
    try {
      frame = ImFrame.fromJson(jsonDecode(raw) as Map<String, dynamic>);
    } catch (_) {
      return;
    }

    final Completer<ImFrame>? waiting = frame.id.isEmpty ? null : _pending.remove(frame.id);
    if (waiting != null) {
      if (!waiting.isCompleted) waiting.complete(frame);
      return;
    }

    if (frame.target == ImPushTarget.kick) {
      _handleKickPush(frame);
      return;
    }

    for (final void Function(ImFrame) listener in List<void Function(ImFrame)>.of(
        _listeners[frame.target] ?? const <void Function(ImFrame)>[])) {
      try {
        listener(frame);
      } catch (_) {
        // A throwing listener must not stop the others or the receive loop.
      }
    }
  }

  /// A kick that arrives as a push rather than as a close reason.
  ///
  /// Both forms carry the same meaning and must get the same response: **stop reconnecting.**
  /// Reconnecting into a kick is a loop, and the loop is invisible from the app — the socket keeps
  /// coming back and keeps being taken away. This path used to emit the event and then reconnect
  /// anyway.
  /// 被踢有两种到达方式，处理必须一致：终止重连。重连进一次踢下线就是死循环。
  void _handleKickPush(ImFrame frame) {
    final Map<String, dynamic>? detail =
        frame.body?.data is Map<String, dynamic> ? frame.body!.data as Map<String, dynamic> : null;
    final String kickReason = detail?['reason']?.toString() ?? '';
    final ImKickReason reason = ImKickReason.parse('im-kick:$kickReason') ?? ImKickReason.unknown;

    // The gateway also sends 1107 as a business code on the kick body; either is enough.
    final bool terminal = reason.terminal || frame.body?.code == ImErrorCode.kickedByOtherDevice;

    if (!_kicks.isClosed) {
      _kicks.add(ImKickEvent(reason, rawReason: kickReason, detail: detail));
    }

    if (!terminal) return;

    _stopped = true;
    _clearTimers();
    unawaited(_socket?.close(1000, 'im-kick:$kickReason') ?? Future<void>.value());
  }

  void _handleClosed(ImSocket socket) {
    if (!identical(socket, _socket)) return;

    _clearTimers();
    _failAllPending();

    if (_stopped) {
      _setState(ImConnectionState.closed);
      return;
    }

    // The close reason is the only thing separating "you were kicked" from "the network died",
    // and the two need opposite responses.
    final ImKickReason? kick = ImKickReason.parse(socket.closeReason);

    if (kick != null) {
      if (!_kicks.isClosed) {
        _kicks.add(ImKickEvent(kick, rawReason: socket.closeReason ?? ''));
      }

      if (kick.terminal) {
        _stopped = true;
        _setState(ImConnectionState.closed);
        return;
      }

      if (kick == ImKickReason.tokenExpired) {
        unawaited(_refreshTokenAndReconnect());
        return;
      }
    }

    _scheduleReconnect();
  }

  void _scheduleReconnect() {
    _setState(ImConnectionState.reconnecting);
    final Duration delay = _backoff.delayFor(++_attempt);
    _reconnectTimer = Timer(delay, () => unawaited(_open()));
  }

  Future<void> _refreshTokenAndReconnect() async {
    final String? next = await _options.onTokenExpired?.call();
    if (next == null || next.isEmpty) {
      _stopped = true;
      _setState(ImConnectionState.closed);
      return;
    }

    _token = next;
    _options = _options.copyWith(token: next);
    _scheduleReconnect();
  }

  void _startHeartbeat() {
    _heartbeatTimer?.cancel();
    _heartbeatTimer = Timer.periodic(_heartbeatInterval, (_) async {
      try {
        final ImHeartbeatResult beat = ImHeartbeatResult.fromJson(
          await request<Map<String, dynamic>>('conn.heartbeat'),
        );

        if (beat.intervalSeconds > 0 && beat.intervalSeconds != _heartbeatInterval.inSeconds) {
          // The server owns the cadence; adopt whatever it reports.
          _heartbeatInterval = Duration(seconds: beat.intervalSeconds);
          _startHeartbeat();
        }
      } catch (_) {
        // A missed beat on a socket the OS still believes is open is exactly the half-open case.
        // Forcing a close is what hands control to the reconnect path.
        unawaited(_socket?.close(4000, 'heartbeat-failed') ?? Future<void>.value());
      }
    });
  }

  void _clearTimers() {
    _heartbeatTimer?.cancel();
    _heartbeatTimer = null;
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
  }

  /// Fails everything still waiting with [ImErrorCode.timeout], **not** with
  /// [ImErrorCode.serviceUnavailable].
  ///
  /// 1005 would claim the call was never delivered, and the SDK does not know that — the request
  /// may well have executed on the server before the socket died. 1004 is the honest answer, and
  /// it is the one that makes a caller reach for `clientMsgId` idempotency instead of blindly
  /// resending.
  /// 套接字带着在途请求断开时报 1004 而不是 1005：请求可能已经在服务端执行了，1005 是在说谎。
  void _failAllPending() {
    if (_pending.isEmpty) return;

    final List<Completer<ImFrame>> waiting = List<Completer<ImFrame>>.of(_pending.values);
    _pending.clear();

    for (final Completer<ImFrame> completer in waiting) {
      if (!completer.isCompleted) {
        completer.completeError(
          const ImException(
            ImErrorCode.timeout,
            'the connection dropped with this request in flight; it may or may not have run',
          ),
        );
      }
    }
  }

  void _setState(ImConnectionState next) {
    if (_state == next) return;
    _state = next;
    if (!_states.isClosed) _states.add(next);
  }
}

File: typescript/src/connection.ts
import { fullJitterDelay } from './backoff.js';
import type { HeartbeatResult } from './models.js';
import { ImError, ImErrorCode, abortError, type ImFrame, type ImRequest } from './protocol.js';
import { SDK_CLIENT_VERSION } from './version.js';

export interface ConnectionOptions {
  /** Base URL, e.g. wss://im.example.com. The channel path is appended. */
  endpoint: string;
  channel?: string;
  appId: string;
  /** User token issued by the tenant backend. Never an app secret. */
  token: string;
  deviceId: string;
  platform: number;
  clientVersion?: string;
  language?: string;
  /** Milliseconds before an unanswered request rejects. */
  requestTimeoutMs?: number;
  /**
   * Supply a token when the current one expires. Return null to stop reconnecting.
   *
   * Called on two paths: a socket closed with `im-kick:TokenExpired`, and a live request answered
   * `1101 TokenExpired` — the second is repaired with `conn.reauth` on the socket that is already
   * open, so a token expiring mid-session costs one round trip instead of a reconnect.
   */
  onTokenExpired?: () => Promise<string | null>;
  /** Injectable for Node/React Native, where WebSocket is not global. */
  webSocketImpl?: typeof WebSocket;
  /**
   * Source of randomness for the reconnect jitter. Injectable only so tests can be deterministic;
   * production has no reason to pass it.
   */
  random?: () => number;
}

/** Trailing options bag on every typed call. Cancellation is the reason it exists. */
export interface ImRequestOptions {
  /**
   * Abandons the reply.
   *
   * It does **not** cancel the server-side effect: a cancelled `msg.send` may well have sent the
   * message. That is what `clientMsgId` idempotency is for — reissue with the same one and the
   * server returns the original result rather than a second message (CONTRACT §7.5).
   */
  signal?: AbortSignal;

  /** Overrides `ConnectionOptions.requestTimeoutMs` for this call. */
  timeoutMs?: number;
}

export type ConnectionState = 'idle' | 'connecting' | 'open' | 'reconnecting' | 'closed';

interface Pending {
  resolve: (frame: ImFrame) => void;
  reject: (error: Error) => void;
  timer: ReturnType<typeof setTimeout>;
  target: string;
  /** Detaches the abort listener; a pending entry that outlived its signal is a leak. */
  release: () => void;
}

/**
 * One multiplexed WebSocket to the gateway, with the reconnect behaviour a production client
 * actually needs.
 *
 * Reconnecting is not an optional extra here. A gateway holds sockets and therefore deploys
 * rarely, but when it does deploy — or when a node fails, or a phone changes network — every
 * client on that node reconnects at once. Without jittered backoff those clients arrive as a
 * synchronised thundering herd against the very service that just came back up. Every delay below
 * is jittered for that reason.
 *
 * 重连不是可选项：网关一次滚动更新会让该节点上所有客户端同时重连。没有带抖动的退避，
 * 这些客户端会变成打向刚恢复的服务的同步洪峰。
 */
export class ImConnection {
  private socket: WebSocket | null = null;
  private state: ConnectionState = 'idle';
  private readonly pending = new Map<string, Pending>();
  private readonly listeners = new Map<string, Set<(frame: ImFrame) => void>>();
  private readonly stateListeners = new Set<(state: ConnectionState) => void>();

  private attempt = 0;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private heartbeatTimer: ReturnType<typeof setInterval> | null = null;
  private heartbeatSeconds = 30;
  private sequence = 0;
  private stopped = false;
  private token: string;

  /** One reauth at a time: twenty requests failing 1101 together must not fetch twenty tokens. */
  private reauthInFlight: Promise<boolean> | null = null;

  /** Reasons that mean "do not come back": reconnecting would fail identically, forever. */
  private static readonly TerminalKicks = new Set([
    'MultiLoginPolicy',
    'TokenRevoked',
    'UserBanned',
    'AppDisabled',
    'AdminKick',
  ]);

  private readonly random: () => number;

  constructor(private readonly options: ConnectionOptions) {
    this.token = options.token;
    this.random = options.random ?? Math.random;
  }

  get currentState(): ConnectionState {
    return this.state;
  }

  /** The token this connection will hand over on its next handshake. Moves on a successful reauth. */
  get currentToken(): string {
    return this.token;
  }

  /** Requests still waiting for a reply. Exposed so a leak is testable rather than inferred. */
  get pendingCount(): number {
    return this.pending.size;
  }

  onState(listener: (state: ConnectionState) => void): () => void {
    this.stateListeners.add(listener);
    return () => this.stateListeners.delete(listener);
  }

  /** Subscribes to a server push target. Returns an unsubscribe function. */
  on(target: string, listener: (frame: ImFrame) => void): () => void {
    let set = this.listeners.get(target);
    if (!set) {
      set = new Set();
      this.listeners.set(target, set);
    }
    set.add(listener);
    return () => set!.delete(listener);
  }

  /** Idempotent: connecting an already-connecting or already-open connection is a no-op. */
  async connect(): Promise<void> {
    this.stopped = false;
    if (this.state === 'open' || this.state === 'connecting' || this.state === 'reconnecting') {
      return;
    }
    await this.open();
  }

  /** Idempotent, and safe to call from a listener. */
  close(): void {
    this.stopped = true;
    this.clearTimers();
    this.socket?.close(1000, 'client-close');
    this.socket = null;
    this.setState('closed');
    this.failAllPending(new ImError(ImErrorCode.Timeout, 'connection closed'));
  }

  /**
   * Sends a request and resolves with its business payload, throwing ImError on a non-zero code.
   *
   * Requests issued while offline reject immediately rather than queueing — a chat client that
   * silently buffers sends produces messages that arrive minutes late into a conversation that has
   * moved on. The application knows whether the message is still worth sending; the SDK does not.
   *
   * Both layers of the envelope are mapped, and never collapsed into one: `status` says whether the
   * call was delivered, `body.code` says what the business rule decided (CONTRACT §7.2).
   */
  async request<T>(target: string, body?: unknown, options?: ImRequestOptions): Promise<T> {
    try {
      return this.unwrap<T>(target, await this.send(target, body, options));
    } catch (error) {
      // A token that expired mid-session is repairable on the socket that is already open: swap it
      // with conn.reauth and reissue once. Only if that fails does the session drop (CONTRACT §7.4).
      if (this.shouldReauth(error, target) && (await this.reauth())) {
        return this.unwrap<T>(target, await this.send(target, body, options));
      }
      throw error;
    }
  }

  private unwrap<T>(target: string, frame: ImFrame): T {
    // status 2 is "this deployment does not have this endpoint" — the exact signal an SDK newer
    // than a private-deployment server produces. 1008 says that; 1002 NotFound would be read as
    // "your group does not exist" and send the integrator looking in the wrong place.
    if (frame.status === 2) {
      throw new ImError(
        ImErrorCode.UnsupportedOperation,
        frame.msg ?? `endpoint not found on this deployment: ${target}`,
        frame.body?.traceId,
        target,
      );
    }

    if (frame.status !== 0) {
      throw new ImError(
        ImErrorCode.InternalError,
        frame.msg ?? 'the endpoint threw',
        frame.body?.traceId,
        target,
      );
    }

    const result = frame.body;
    if (!result || result.code !== 0) {
      throw new ImError(
        result?.code ?? ImErrorCode.InternalError,
        result?.message ?? 'request failed',
        result?.traceId,
        target,
      );
    }

    return result.data as T;
  }

  private shouldReauth(error: unknown, target: string): boolean {
    return (
      error instanceof ImError &&
      error.code === ImErrorCode.TokenExpired &&
      target !== 'conn.reauth' &&
      this.options.onTokenExpired !== undefined
    );
  }

  /** Swaps the token on the live socket. Shared, so concurrent 1101s cost one token fetch. */
  private reauth(): Promise<boolean> {
    this.reauthInFlight ??= (async () => {
      try {
        const next = await this.options.onTokenExpired?.();
        if (!next) return false;

        this.unwrap<void>('conn.reauth', await this.send('conn.reauth', { token: next }));
        this.token = next;
        return true;
      } catch {
        // The session drops the ordinary way: the server closes the socket, or the next request
        // fails again and the caller sees 1101 with requiresReauth set.
        return false;
      } finally {
        this.reauthInFlight = null;
      }
    })();

    return this.reauthInFlight;
  }

  private send(target: string, body?: unknown, options?: ImRequestOptions): Promise<ImFrame> {
    if (options?.signal?.aborted) {
      return Promise.reject(abortError(`cancelled before sending: ${target}`));
    }

    if (!this.socket || this.state !== 'open') {
      return Promise.reject(new ImError(ImErrorCode.ServiceUnavailable, 'not connected', null, target));
    }

    const id = `c${++this.sequence}-${Date.now().toString(36)}`;
    const payload: ImRequest = { id, target, body: body ?? {} };
    const timeoutMs = options?.timeoutMs ?? this.options.requestTimeoutMs ?? 15000;

    return new Promise<ImFrame>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.settle(id);
        reject(new ImError(ImErrorCode.Timeout, `request timed out: ${target}`, null, target));
      }, timeoutMs);

      const signal = options?.signal;
      const onAbort = () => {
        // Removing the entry is the whole job here: a cancellation path that leaves it behind is a
        // leak that grows for the life of the connection.
        this.settle(id);
        reject(abortError(`cancelled: ${target}`));
      };

      signal?.addEventListener('abort', onAbort, { once: true });

      this.pending.set(id, {
        resolve,
        reject,
        timer,
        target,
        release: () => signal?.removeEventListener('abort', onAbort),
      });

      try {
        this.socket!.send(JSON.stringify(payload));
      } catch (error) {
        this.settle(id);
        reject(error instanceof Error ? error : new Error(String(error)));
      }
    });
  }

  /** Clears one pending entry and everything hanging off it. */
  private settle(id: string): Pending | undefined {
    const entry = this.pending.get(id);
    if (!entry) return undefined;

    clearTimeout(entry.timer);
    entry.release();
    this.pending.delete(id);
    return entry;
  }

  private async open(): Promise<void> {
    this.setState(this.attempt === 0 ? 'connecting' : 'reconnecting');

    const Impl = this.options.webSocketImpl ?? globalThis.WebSocket;
    if (!Impl) {
      throw new Error('No WebSocket implementation available. Pass options.webSocketImpl.');
    }

    const socket = new Impl(this.buildUrl());
    this.socket = socket;

    socket.onopen = () => {
      this.attempt = 0;
      this.setState('open');
      this.startHeartbeat();
    };

    socket.onmessage = (event: MessageEvent) => this.dispatch(event.data);

    socket.onclose = (event: CloseEvent) => {
      this.clearTimers();

      // 1004 Timeout, not 1005 ServiceUnavailable: 1005 claims the call was never delivered, and
      // the SDK does not know that — the request may well have executed. 1004 is the honest
      // answer, and it is the one that makes a caller reach for clientMsgId idempotency instead of
      // blindly resending (CONTRACT §7.2).
      this.failAllPending(new ImError(ImErrorCode.Timeout, 'connection lost with the request in flight'));

      if (this.stopped) {
        this.setState('closed');
        return;
      }

      // The close reason is the only signal that separates "you were kicked" from "the network
      // died", and the two need opposite responses.
      const kick = this.parseKick(event.reason);
      if (kick && ImConnection.TerminalKicks.has(kick)) {
        this.emitKick(kick);
        this.setState('closed');
        return;
      }

      if (kick === 'TokenExpired') {
        void this.refreshTokenAndReconnect();
        return;
      }

      this.scheduleReconnect();
    };

    socket.onerror = () => {
      // onclose always follows; reconnect logic lives there so it runs exactly once.
    };
  }

  private buildUrl(): string {
    const base = this.options.endpoint.replace(/\/+$/, '');
    const channel = this.options.channel ?? '/im';
    const params = new URLSearchParams({
      appId: this.options.appId,
      token: this.token,
      deviceId: this.options.deviceId,
      platform: String(this.options.platform),
      v: '1',
    });

    // `cv` defaults to this SDK's own version so a support ticket carries it without anyone having
    // to ask. An application that wants to report its own build overrides it.
    params.set('cv', this.options.clientVersion || SDK_CLIENT_VERSION);
    if (this.options.language) params.set('lang', this.options.language);

    return `${base}${channel}?${params.toString()}`;
  }

  private dispatch(raw: unknown): void {
    if (typeof raw !== 'string') {
      return;
    }

    let frame: ImFrame;
    try {
      frame = JSON.parse(raw) as ImFrame;
    } catch {
      return;
    }

    const waiting = frame.id ? this.settle(frame.id) : undefined;
    if (waiting) {
      waiting.resolve(frame);
      return;
    }

    const listeners = this.listeners.get(frame.target);
    if (listeners) {
      for (const listener of listeners) {
        try {
          listener(frame);
        } catch {
          // A throwing listener must not stop the others or the receive loop.
        }
      }
    }
  }

  private scheduleReconnect(): void {
    this.setState('reconnecting');
    this.attempt += 1;

    // Full jitter — a uniform draw from 0..ceiling, not the ceiling and not ceiling±10%. See
    // backoff.ts for why that distinction is the difference between a ramp and a second outage.
    const delay = fullJitterDelay(this.attempt, this.random);

    this.reconnectTimer = setTimeout(() => void this.open(), delay);
  }

  private async refreshTokenAndReconnect(): Promise<void> {
    const next = await this.options.onTokenExpired?.();
    if (!next) {
      this.setState('closed');
      this.emitKick('TokenExpired');
      return;
    }

    this.token = next;
    this.scheduleReconnect();
  }

  private startHeartbeat(): void {
    this.stopHeartbeat();
    this.heartbeatTimer = setInterval(() => {
      this.request<HeartbeatResult>('conn.heartbeat')
        .then((result) => {
          // The server owns the cadence; adopt whatever it reports.
          if (result.intervalSeconds > 0 && result.intervalSeconds !== this.heartbeatSeconds) {
            this.heartbeatSeconds = result.intervalSeconds;
            this.startHeartbeat();
          }
        })
        .catch(() => {
          // A missed beat on a socket the OS still thinks is open is exactly the half-open
          // connection case: force a close so reconnect logic takes over.
          this.socket?.close(4000, 'heartbeat-failed');
        });
    }, this.heartbeatSeconds * 1000);
  }

  private stopHeartbeat(): void {
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
  }

  private clearTimers(): void {
    this.stopHeartbeat();
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
  }

  private failAllPending(error: Error): void {
    for (const id of [...this.pending.keys()]) {
      const entry = this.settle(id);
      entry?.reject(error);
    }
    this.pending.clear();
  }

  private parseKick(reason: string | undefined): string | null {
    if (!reason || !reason.startsWith('im-kick:')) {
      return null;
    }
    return reason.slice('im-kick:'.length);
  }

  private emitKick(reason: string): void {
    const listeners = this.listeners.get('conn.kick');
    if (!listeners) return;

    const frame: ImFrame = {
      id: '',
      target: 'conn.kick',
      status: 0,
      body: { code: 0, serverTime: Date.now(), data: { reason } },
    };

    for (const listener of listeners) {
      try {
        listener(frame);
      } catch {
        /* ignore */
      }
    }
  }

  private setState(state: ConnectionState): void {
    if (this.state === state) return;
    this.state = state;
    for (const listener of this.stateListeners) {
      try {
        listener(state);
      } catch {
        /* ignore */
      }
    }
  }
}
